core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -524,49 +524,26 @@ private[master] class Master(
   }
 
   /**
-   * Can an app use the given worker? True if the worker has enough memory and we haven't already
-   * launched an executor for the app on it (right now the standalone backend doesn't like having
-   * two executors on the same worker).
+   * Can an app use the given worker?
    */
   private def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+    val enoughResources = worker.memoryFree >= app.desc.memoryPerExecutorMB && worker.coresFree > 0
+    val allowToExecute = app.desc.coresPerExecutor.isDefined || !worker.hasExecutor(app)
+    allowToExecute && enoughResources
   }
 
   /**
-   * Schedule the currently available resources among waiting apps. This method will be called
-   * every time a new app joins or resource availability changes.
+   * The resource allocator spread out each app among all the workers until it has all its cores in
+   * spreadOut mode otherwise packs each app into as few workers as possible until it has assigned
+   * all its cores. User can define spark.deploy.maxCoresPerExecutor per application to
+   * limit the maximum number of cores to allocate to each executor on each worker; if the parameter
+   * is not defined, then only one executor will be launched on a worker.
```

Comment:
This java doc is verbose and outdated. Can you do the following instead:
```
Schedule executors to be launched on the workers.

There are two modes of launching executors. The first attempts to spread out
an application's executors on as many workers as possible, while the second
does the opposite (i.e. as few workers as possible). The former is usually
better for data locality purposes and is the default.

The number of cores assigned to each executor is configurable. When this
is explicitly set, multiple executors from the same application may be launched
on the worker if the worker has enough cores. Otherwise, each executor grabs
all the cores available on the worker by default, in which case only one executor
per application may be launched on the worker.
```